odoo/addons/base/tests/test_base.py

```diff
@@ -48,17 +61,250 @@ def test_04_literal_eval_forbidden(self):
+            safe_eval("(0 for foo.bar in range(10))", globals_dict={"foo": Foo()}, sandboxed_instances=(Foo, ), mode="exec")
+
+        safe_eval("for (a, b) in enumerate(range(10)): ...", mode="exec")
+        safe_eval("(a for (a, b) in enumerate(range(10)))", mode="exec")
+
+        with self.assertRaisesRegex(SyntaxError, "Nodes of type 'Delete' are not allowed."):
+            safe_eval("del foo.bar", globals_dict={"foo": Foo()}, sandboxed_instances=(Foo, ), mode="exec")
+
+    def test_12_deny_dunder_arg(self):
+        with self.assertRaisesRegex(NameError, "Forbidden name '__test'"):
+            safe_eval("def f(x, *, __test): ...", mode="exec")
+
+        with self.assertRaisesRegex(NameError, "Forbidden name '__test'"):
+            safe_eval("def f(x, /, __test): ...", mode="exec")
+
+        with self.assertRaisesRegex(NameError, "Forbidden name '__test'"):
+            safe_eval("def f(x, *__test): ...", mode="exec")
+
+    def test_13_deny_dangerous_func(self):
+        with self.assertRaisesRegex(ValueError, r"You cannot pass <built-in function eval> \(method of <module 'builtins' \(built-in\)>\) to any objects."):
+            safe_eval("eval('1')", mode="exec", globals_dict={"eval": eval})
+
+        with self.assertRaisesRegex(ValueError, r"You cannot pass <built-in function exec> \(method of <module 'builtins' \(built-in\)>\) to any objects."):
+            safe_eval("exec('1')", mode="exec", globals_dict={"exec": exec})
+
+        with self.assertRaisesRegex(ValueError, r"You cannot pass <function safe_eval at *"):
+            safe_eval("safe_eval('1')", mode="exec", globals_dict={"safe_eval": safe_eval})
+
+        with self.assertRaisesRegex(ValueError, r"You cannot pass <built-in function system> \(method of <module '*'"):
+            safe_eval("system('echo')", mode="exec", globals_dict={"system": os.system})
+
+        safe_eval("time.strftime('%Y-%m-%d')", mode="exec", globals_dict={"time": time})
+
+    def test_14_deny_use_kwarg(self):
+        with self.assertRaisesRegex(SyntaxError, "Forbidden use of kwargs."):
+            safe_eval("def f(x, **kwargs): ...", mode="exec")
+
+        with self.assertRaisesRegex(SyntaxError, "Forbidden use of kwargs."):
+            safe_eval("lambda x, **kwargs: ...", mode="exec")
+
+    def test_15_deny_use_decorator(self):
+        with self.assertRaisesRegex(SyntaxError, "Forbidden use of decorators."):
+            safe_eval(cleandoc(
+                """
+                @foo
+                def f(x):
+                    ...
+                """
+            ))
+
+    def test_16_deny_use_dunder_exception(self):
+        with self.assertRaisesRegex(NameError, "Forbidden name '__type_checker'"):
+            safe_eval(cleandoc(
+                """
+                try:
+                    ...
+                except Exception as __type_checker:
+                    ...
+                """
+            ))
+
+    def test_17_globals_dict_leak_builtins(self):
```

Comment:
can we do the same the the locals ?

Reply:
Should I just check for the presence of dunders inside of the local context ?

Reply:
yes, if you manage to test that (maybe by passing a function that manually change `__locals__`, to see if it's possible)